### PR TITLE
added application-lib

### PR DESCRIPTION
**Pull request by bisztray**

### CODE REVIEW

File: application-lib/pkg/handlers/handlers.go [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+func deleteChangedResources(pna k8sdynamic2.ResourceDescriptor) error {
+	k8sClient := k8sdynamic2.New(kubelib.GetKubeAPI())
+	if err := k8sClient.DeleteResources([]k8sdynamic2.ResourceDescriptor{pna}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func waitUntilResourcesAreReleased(pna k8sdynamic2.ResourceDescriptor) {
```
Comment (howardhalili) [inline application-lib/pkg/handlers/handlers.go:192]: we may need (waiting) timeout in verifying resource deletion.

File: application-lib/pkg/types/common_types.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+type AppStatus string
+
+const (
+	AppStatusNotSet     = "UNSET"
+	AppStatusNotRunning = "NOT_RUNNING"
+	AppStatusRunning    = "RUNNING"
+	AppStatusFrozen     = "FROZEN"
+)
+
+type PrivateNetworkAccess struct {
```
Comment (howardhalili) [inline application-lib/pkg/types/common_types.go:16]: is it possible to remove the old(/to be deprecated) fields?

File: application-lib/pkg/config/config.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+// SPDX-License-Identifier: BSD-3-Clause
+
+package config
+
+import (
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"os"
+)
+
+const operatorConfigFilename = "operatorconfig.yaml"
```
Comment (hochdorf) [inline application-lib/pkg/config/config.go:12]: why do we need this yaml file here? Shouldn't it be part of the generator?

File: application-lib/pkg/handlers/handlers.go [hunk 4]
```diff
@@ -0,0 +1,433 @@
+package handlers
+
+// Copyright 2022 Nokia
+// Licensed under the BSD 3-Clause License.
+// SPDX-License-Identifier: BSD-3-Clause
+
+import (
+	"context"
+	"encoding/json"
+	k8sdynamic2 "github.com/nokia/industrial-application-framework/application-lib/pkg/k8sdynamic"
```
Comment (hochdorf) [inline application-lib/pkg/handlers/handlers.go:10]: can you pls rename this import to just k8sdynamic?

File: application-lib/README.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# Kubernetes application library
```
Comment (hochdorf) [inline application-lib/README.md:1]: would be useful to include the relevant parts of the consul-operator readme here

File: application-lib/pkg/monitoring/monitoring.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
+// Copyright 2020 Nokia
+// Licensed under the BSD 3-Clause License.
+// SPDX-License-Identifier: BSD-3-Clause
+
+package monitoring
+
+import (
+	"context"
+	kubelib2 "github.com/nokia/industrial-application-framework/application-lib/pkg/kubelib"
```
Comment (hochdorf) [inline application-lib/pkg/monitoring/monitoring.go:9]: pls rename to kubelib

File: application-lib/pkg/types/common_interfaces.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	GetAppStatus() AppStatus
+	GetAppliedResources() []k8sdynamic.ResourceDescriptor
+	SetAppliedResources(resources []k8sdynamic.ResourceDescriptor)
+	GetPrevSpec() OperatorSpec
+	GetPrevSpecDeepCopy() OperatorSpec
+	SetPrevSpec(spec OperatorSpec) error
+	GetAppReportedData() CommonAppReportedData
+}
+
+type CommonAppReportedData interface {
```
Comment (hochdorf) [inline application-lib/pkg/types/common_interfaces.go:37]: do we need this type? as i remember we discussed that it is fully app specific.

File: application-lib/pkg/types/common_types.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+type Network struct {
+	ApnUUID          string   `json:"apnUUID,omitempty"`
+	NetworkID        string   `json:"networkId,omitempty"`
+	AdditionalRoutes []string `json:"additionalRoutes,omitempty"`
+}
+
+type AppPodFixIp struct {
+	Db string `json:"db"`
```
Comment (hochdorf) [inline application-lib/pkg/types/common_types.go:32]: this is consul specific, shouldn't be the part of the lib

**Review by hochdorf — CHANGES_REQUESTED** [review]

File: application-lib/pkg/config/config.go (hunk 3 above)
Comment (bisztray) [reply application-lib/pkg/config/config.go:12]: This yaml contains runtime information on the operator. It is not part of the generation.

File: application-lib/pkg/config/config.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+type TemplateConfig struct {
+	LeftDelimiter  string `yaml:"leftDelimiter"`
+	RightDelimiter string `yaml:"rightDelimiter"`
+}
+
+func GetConfiguration(configDir string) (OperatorConfig, error) {
+	operatorConfigYaml := configDir + "/" + operatorConfigFilename
+
+	var err error
+	data, err := os.ReadFile(operatorConfigYaml)
```
[the remark this answers is not in the document]
Comment (bisztray) [reply application-lib/pkg/config/config.go]: fixedf

File: application-lib/pkg/monitoring/monitoring.go [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+func (m *Monitor) Pause() {
+	if m.Running {
+		log.Info("Watching application paused")
+		m.Running = false
+		close(m.pauseChannel)
+	}
+}
+
+func (m *Monitor) GetApplicationStatus() common_types.AppStatus {
```
[the remark this answers is not in the document]
Comment (bisztray) [reply application-lib/pkg/monitoring/monitoring.go:153]: Fixing this properly is a new feature in itself, updated in an easier manner (not dirty, but not as elaborate as on other examples)

File: application-lib/pkg/template/template.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func (t *Templater) runGoTemplate(joinSeparator string) (string, error) {
+	log.Info("Running go templates")
+
+	return t.templateDir(t.WorkDir, joinSeparator)
+}
+
+func (t *Templater) templateFile(workDir string, file os.FileInfo, joinSeparator string) (string, error) {
+	if strings.Contains(file.Name(), "yaml") {
```
[the remark this answers is not in the document]
Comment (bisztray) [reply application-lib/pkg/template/template.go]: done

File: application-lib/pkg/types/common_types.go (hunk 2 above)
Comment (bisztray) [reply application-lib/pkg/types/common_types.go]: done

File: application-lib/pkg/handlers/handlers.go (hunk 4 above)
Comment (bisztray) [reply application-lib/pkg/handlers/handlers.go]: done

File: application-lib/pkg/monitoring/monitoring.go (hunk 6 above)
Comment (bisztray) [reply application-lib/pkg/monitoring/monitoring.go:9]: done

File: application-lib/pkg/types/common_interfaces.go (hunk 7 above)
Comment (bisztray) [reply application-lib/pkg/types/common_interfaces.go]: removed, and moved back to to app-specific types

File: application-lib/pkg/monitoring/monitoring.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (howardhalili) [reply application-lib/pkg/monitoring/monitoring.go:153]: looks fine given where it is called

**Review by howardhalili — APPROVED** [review]

File: application-lib/pkg/monitoring/monitoring.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (howardhalili) [reply application-lib/pkg/monitoring/monitoring.go:153]: something came out on top of my head, i may not need to rescind my approval but will you open a git issue about this once this is merge?
scenario such as miscalculated resourcerequest/quota (smaller than actually needed) may end up reporting false running status. monitoring is running in this case.

[the remark this answers is not in the document]
Comment (bisztray) [reply application-lib/pkg/monitoring/monitoring.go:153]: ok

File: application-lib/pkg/monitoring/monitoring.go (hunk 6 above)
Comment (hochdorf) [reply application-lib/pkg/monitoring/monitoring.go:9]: it is still kubelib2 not just kubelib

Comment (bisztray) [reply application-lib/pkg/monitoring/monitoring.go]: I hope this is done now

**Review by hochdorf — APPROVED** [review]

**Pull request merged**